File: cpp/Flatten_Binary_Tree_to_Linked_List.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *visited;
    void flatten(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (root == NULL) return;
        solve(root);
    }
    void solve(TreeNode *root){
        visited = root;
        if (root->left == NULL && root->right == NULL) return;
        if (root->left != NULL){                // flatten the left side
            solve(root->left);
            visited->right = root->right;
            root->right = root->left;           // move the flattered left tree to the right
            root->left = NULL;
        }
        if (visited->right != NULL){            // flatten the right sight
            solve(visited->right);
        }
    }
};

File: java/Flatten_Binary_Tree_to_Linked_List.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode visited;
    public void flatten(TreeNode root) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if (root == null) return;
        visited = null;
        solve (root);
    }
    private void solve(TreeNode root) {
        visited = root;
        if (root.left == null && root.right == null) return;
        if (root.left != null) {
            solve(root.left);
            visited.right = root.right;
            root.right = root.left;
            root.left = null;
        }
        if (visited.right != null) {
            solve (visited.right);
        }
    }
}
